### PR TITLE
history data read, asynchronous user validation, event removeMonitoredItem

### PR DESCRIPTION
**Pull request by hr83**

### CODE REVIEW

**Comment by erossignon** [comment]
Great job!
It would be nice to get a set of unit test to stress the new code so I can merge it to the master branch.
Let me know if you need help to get on with writing the test.

**Comment by erossignon** [comment]
Great job!
It would be nice to get a set of unit test to stress the new code so I can merge it to the master branch.
Let me know if you need help to get on with writing the test.

**Comment by hr83** [comment]
To be honest, I've never done that before in a professional manner. It will take some time for me to write a test. I will come back to you if I need help.

**Comment by hr83** [comment]
Etienne, how can I contact you outside of this pull request?

**Comment by erossignon** [comment]
try this command : ```git --no-pager show -s --format="%aE" 845c924````  to discover how

**Comment by hr83** [comment]
I have added a few other features now and created test cases for them.

File: lib/address_space/ua_variable.js [hunk 1]
```diff
@@ -499,6 +502,26 @@ UAVariable.prototype.readAttribute = function (attributeId, indexRange, dataEnco
...
+ * @param continuationPoint {ByteString}
+ * @param callback {Function}
+ * @param callback.err
+ * @param callback.result {HistoryReadResult}
+ */
+UAVariable.prototype.historyRead = function (historyReadDetails, indexRange, dataEncoding, continuationPoint, callback) {
+    if (continuationPoint) {
+        //TODO: use continutaion point manager to support continuation of historical reads
+        callback(null, new HistoryReadResult({statusCode: StatusCodes.BadNotSupported}));
+    } else if (typeof this["_historyRead"] != "function") {
```
Comment (erossignon) [inline lib/address_space/ua_variable.js:519]: _.isFunction(this._historyRead) ?

File: lib/server/opcua_server.js [hunk 2]
```diff
@@ -893,7 +903,7 @@ OPCUAServer.prototype.isValidUserIdentityToken = function (channel, session, use
     }
     return true;
 };
-OPCUAServer.prototype.isUserAuthorized = function (channel, session, userIdentityToken) {
+OPCUAServer.prototype.isUserAuthorized = function (channel, session, userIdentityToken, done) {
 
     var self = this;
```
Comment (erossignon) [inline lib/server/opcua_server.js:908]: - 'done' doesn't comply to the  `callback(err,data)` 
- assert(_.isFunction(done));

File: lib/server/opcua_server.js [hunk 3]
```diff
@@ -1058,78 +1068,79 @@ OPCUAServer.prototype._on_ActivateSessionRequest = function (message, channel) {
...
-                arrayType: VariantArrayType.Array,
-                value: []
-            },
-            // UserIdentityToken reflects the userIdentityToken parameter of the ActivateSession Service call.
-            // For Username/Password tokens the password should NOT be included.
-            userIdentityToken: {
-                dataType: "ExtensionObject" /*  UserIdentityToken */,
-                value: userIdentityTokenPasswordRemoved(session.userIdentityToken)
-            },
+    server.isUserAuthorized(channel, session, request.userIdentityToken, function(authorized){
```
Comment (erossignon) [inline lib/server/opcua_server.js:1071]:
``` javascript
server.isUserAuthorized(channel, session, request.userIdentityToken, function(err,authorized){
    if (err) { return rejectConnection(StatusCodes.BadInternalError); }
    if (!authorized) {
    }
```

File: lib/server/opcua_server.js [hunk 4]
```diff
@@ -1058,78 +1068,79 @@ OPCUAServer.prototype._on_ActivateSessionRequest = function (message, channel) {
...
+            session.nonce = server.makeServerNonce();
+
+            session.status = "active";
+
+            response = new ActivateSessionResponse({serverNonce: session.nonce});
+            channel.send_response("MSG", response, message);
 
-            // SecureChannelId shall uniquely identify the SecureChannel. The application shall use the same identifier
-            // in all AuditEvents related to the Session Service Set (AuditCreateSessionEventType,
-            // AuditActivateSessionEventType and their subtypes) and the SecureChannel Service Set
```
Comment (erossignon) [inline lib/server/opcua_server.js:1126]: why this comment has been removed ?

File: lib/server/opcua_server.js [hunk 5]
```diff
@@ -1058,78 +1068,79 @@ OPCUAServer.prototype._on_ActivateSessionRequest = function (message, channel) {
...
-            // (AuditChannelEventType and its subtypes).
-            secureChannelId: {dataType: "String", value: session.channel.secureChannelId.toString()}
+            var userIdentityTokenPasswordRemoved=function(userIdentityToken) {
+                var a = userIdentityToken;
+                // to do remove password
```
Comment (erossignon) [inline lib/server/opcua_server.js]: What is the purpose of this ?

File: pki/untrusted/76ced20dc060e04abfed723f32151e4b376a1fd8.pem [hunk 6]
```diff
@@ -0,0 +1,24 @@
+-----BEGIN CERTIFICATE-----
+MIID9TCCAt2gAwIBAgICEAQwDQYJKoZIhvcNAQELBQAwWjELMAkGA1UEBhMCRlIx
```
Comment (erossignon) [inline pki/untrusted/76ced20dc060e04abfed723f32151e4b376a1fd8.pem:2]: remove this file

File: pki/untrusted/8d24262c6821d751032260e257fa071a1879d3f7.pem [hunk 7]
```diff
@@ -0,0 +1,24 @@
+-----BEGIN CERTIFICATE-----
```
Comment (erossignon) [inline pki/untrusted/8d24262c6821d751032260e257fa071a1879d3f7.pem:1]: remove this file

File: lib/address_space/ua_variable.js (hunk 1 above)
Comment (hr83) [inline lib/address_space/ua_variable.js:519]: sure, should use underscore here. It's the only thing that you did not fix - could you do that too?

File: lib/server/opcua_server.js (hunk 4 above)
Comment (hr83) [inline lib/server/opcua_server.js:1126]: was not intended

**Pull request merged**